craft_application/services/project.py

```diff
@@ -0,0 +1,349 @@
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""A service for handling access to the project."""
+
+from __future__ import annotations
+
+import copy
+import os
+import pathlib
+from typing import TYPE_CHECKING, Any, cast, final
+
+import craft_parts
+import craft_platforms
+from craft_cli import emit
+
+from craft_application import errors, grammar, util
+from craft_application.models.grammar import GrammarAwareProject
+
+from . import base
+
+if TYPE_CHECKING:
+    from craft_application import models
+    from craft_application.application import AppMetadata
+
+    from .service_factory import ServiceFactory
+
+
+class ProjectService(base.AppService):
+    """A service for handling access to the project."""
+
+    __platforms: dict[str, craft_platforms.PlatformDict] | None
+    __project_file_path: pathlib.Path | None
+    __raw_project: dict[str, Any] | None
+    _project_dir: pathlib.Path
+    _project_model: models.Project | None
+
+    def __init__(
+        self, app: AppMetadata, services: ServiceFactory, *, project_dir: pathlib.Path
+    ) -> None:
+        super().__init__(app, services)
+        self.__platforms = None
+        self.__project_file_path = None
+        self.__raw_project: dict[str, Any] | None = None
+        self._project_dir = project_dir
+        self._project_model = None
+
+    def resolve_project_file_path(self) -> pathlib.Path:
+        """Get the path to the project file from the root project directory.
+
+        The default behaviour is to find the project file directly in the directory
+        based on the app name or raise an exception. However, an application may
+        override this if necessary. For example, Snapcraft needs to override this to
+        check other possible directories.
+
+        :param project_dir: The base project directory to search.
+        :returns: The path to the extant project file
+        :raises: ProjectFileMissingError if the project file could not be found.
+        """
+        if self.__project_file_path:
+            return self.__project_file_path
+        try:
+            path = (self._project_dir / f"{self._app.name}.yaml").resolve(strict=True)
+        except FileNotFoundError as err:
+            raise errors.ProjectFileMissingError(
+                f"Project file '{self._app.name}.yaml' not found in '{self._project_dir}'.",
+                details="The project file could not be found.",
+                resolution="Ensure the project file exists.",
+                retcode=os.EX_NOINPUT,
+            ) from err
+        emit.trace(f"Project file found at {path}")
+        return path
+
+    def _load_raw_project(self) -> dict[str, Any]:
```

Comment:
Should this method be final?

Reply:
I was debating it and decided not to "just in case," but your raising this made me reconsider since it's much easier to make it not final later than it is to make it final.

In light of that, I've made all of the methods I don't intend people to override final